File: FlatternBTtoLL.py
#114. Flatten Binary Tree to Linked List
"""
Time Complexity : O(n)
Space Complexity : O(height of tree)
"""
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    result = []
    def helper(self, root):
        #base case
        if root == None:
            return
        
        #logic
        if root.left != None:
            self.helper(root.left)
            temp = root.right #storing right because we will be loosing it
            root.right = root.left
            root.left = None #no need for left node now. None required acc. question
            while root.right != None: #go till end of the flattern temp LL 
                                    #so that temp can be atttached at end.
                root = root.right 
                
            root.right = temp
        
        if root.right != None:
            self.helper(root.right)
    
    def flatten(self, root: Optional[TreeNode]) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        #null case
        if root == None:
            return None
        
        self.helper(root)
        

File: flatternBTtoLL.java
// Time Complexity : O(N) 
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this: No

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public void flatten(TreeNode root) {
        //base
        if(root == null || root.left == null && root.right == null)return;
        //recursive case
        if(root.left != null){
            flatten(root.left);
            TreeNode tempright = root.right;
            root.right = root.left;
            root.left = null;
            while(root.right != null) root = root.right;
            root.right = tempright;
        }
        flatten(root.right);
    }
}